process: Send signal name to signal handlers

robertrossmann opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
process

##### Description

Allow the signal handler to receive the signal code that triggered the handler.

This is useful in situations where you have one common signal handler to gracefully exit a Node.js process (ie. one handler for `SIGINT` and `SIGTERM`) **and** you still want to know exactly which signal was used to stop the process.


###### Before

```js
function shutdown(signal) {
  // Clean up resources
  console.log(`Quit - received ${signal}`)
}

process.once('SIGTERM', () => shutdown('SIGTERM'))
process.once('SIGINT', () => shutdown('SIGINT'))
```

###### After

```js
function shutdown(signal) {
  // Clean up resources
  console.log(`Quit - received ${signal}`)
}

process.once('SIGTERM', shutdown)
process.once('SIGINT', shutdown)
```

robertrossmann commented:
@bnoordhuis Thanks for the feedback - I have swapped the arguments as requested.

One additional reason why this could be beneficial is that currently if you want to know the signal's code you have to look it up in the `os.constants.signals` table. Having the code an argument of the signal handler simplifies this process.

evanlucas commented:
@robertrossmann thanks for taking the time to put together this PR. We have something similar in most of our services, so I can definitely appreciate the value in this. I'd be happy to help out to try and get the test working (at least on macOS and linux) as well

robertrossmann commented:
Thanks @evanlucas for offering to help me with the tests. To describe current situation:

The test attaches a signal listener wrapped in `common.mustCall()` and in that listener function the arguments are checked to match expectations. Then, a signal is sent to the same process using `process.kill(process.pid, 'SIGINT)`. However, the `common.mustCall()` wrapper throws an error stating that the listener was not called.

I wrote most of the test code by following [test-signal-handler.js](https://github.com/robertrossmann/node/blob/process-signal-args/test/parallel/test-signal-handler.js) test. I am unsure if the new test suite needs anything extra.

cjihrig commented on test/parallel/test-signal-args.js:
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+const assert = require('assert');
+const common = require('../common');
```
Please move this up a line so that it is the first `require()`.

cjihrig commented on test/parallel/test-signal-args.js:
```diff
@@ -0,0 +1,21 @@
+const common = require('../common');
+const os = require('os');
+
+console.log(`process.pid: ${process.pid}`);
```
Please drop the `console.log()`.

cjihrig commented on test/parallel/test-signal-args.js:
```diff
@@ -0,0 +1,21 @@
+process.once('SIGINT', common.mustCall((signal, code) => {
+  assert.strictEqual(code, os.constants.signals[signal]);
+  assert.strictEqual(signal, 'SIGINT');
+}), 1);
```
You can drop the `1` here and below, as that is the default value.

cjihrig commented on doc/api/process.md:
```diff
@@ -362,6 +362,14 @@ process.stdin.resume();
+process.on('SIGINT', handle);
+process.on('SIGTERM', handle);
+
+function handle(signal, code) {
```
I think the documentation needs to mention that these new arguments are provided.

BridgeAR replied:
I am sorry but why is the signal added as well? The handler itself already listens to the signal and that is redundant therefore?

robertrossmann commented:
@cjihrig Thank you for your feedback - I have updated the code as requested and added extra sentence to the docs about the function arguments.

cjihrig commented on doc/api/process.md:
```diff
@@ -350,6 +350,9 @@ Signal events will be emitted when the Node.js process receives a signal. Please
 `SIGINT`, `SIGHUP`, etc.
 
+The signal handler will receive two arguments: the signal's name (`'SIGINT'`,
+ `'SIGTERM'` etc.) and the signal's numeric code, (`2`, `15`, respectively).
```
Missing a comma after `SIGTERM`.

numeric code, (`2`, `15`, respectively) -> numeric code, respectively.

robertrossmann replied:
Fixed, thanks. I preserved the numeric examples, though - let me know if you intended to have them removed.

cjihrig replied:
I did mean to remove them. I don't think they add much value. If you're going to keep them, please throw an `, etc.` in there too.

Fishrock123 commented:
Sorry for being pedantic but, can't this just be put in a module that maps separate handlers with their appropriate info and then does the dispatch?

If that already exists, does it not work well?

robertrossmann commented:
@Fishrock123 You are correct that the functionality included in this PR can easily be achieved by other means (see my initial example). It is very easy to wrap a signal handler in an arrow function and pass the signal name to the actual handler, and it is equally easy to look up that signal's code from `os.constants.signals`.

This PR is just an attempt to improve the ergonomics of such task by not requiring the developer to do these extra steps when Node.js already can provide all the information necessary.

So, yes - Node.js does not _need_ this - it does not introduce any new features or functionality. It just attempts to improve the way people work with signal handlers.

BridgeAR commented:
I am -0 on this if this should land I would prefer to land it without the redundancy of the signal.

robertrossmann commented:
@BridgeAR The signal is given to the handler so that the handler knows **which signal** triggered it. This is useful in situations where the same handler is listening on multiple signals and you **want to know** which signal caused the handler to be triggered.

As I mentioned earlier, the whole point of this PR is merely **ergonomics** - I can still achieve my goal using other means (arrow functions, looking up the signal code elsewhere etc.).

bnoordhuis commented on doc/api/process.md:
```diff
@@ -350,6 +350,9 @@ Signal events will be emitted when the Node.js process receives a signal. Please
 `SIGINT`, `SIGHUP`, etc.
 
+The signal handler will receive two arguments: the signal's name (`'SIGINT'`,
+ `'SIGTERM'`, etc.) and the signal's numeric code.
```
Maybe move this below the next paragraph or merge them.  I think an unwitting reader would assume the second one is a continuation of the first, i.e., that it's about the callback's arguments.

Fishrock123 commented:
I think I'm ok with this, but I do also prefer @bnoordhuis's recommendation to not pass the signal number.

If the signal number is necessary, it can be retrieved from `os.constants.signals`.

robertrossmann commented:
I have rebased on current master and hopefully addressed all review feedback. I have also fixed the tests 🎉:

https://github.com/nodejs/node/pull/15606/files#diff-05a7b1950b62d59da8caf56feafde636R20

robertrossmann commented:
It seems the test suite got stuck somewhere. 🤔 Can someone look at it and perhaps restart it? Thank you!

PS. Should I rebase once again or is it okay if this is based off of some older commit?

gireeshpunathil commented:
new CI: https://ci.nodejs.org/job/node-test-pull-request/10749/

robertrossmann commented:
Ooops, looks like I forgot to run `make lint` before pushing and there were 2 linter errors being reported. Not sure why the linter CI run did not report the build as failed and instead stalled - when I looked at the job report there were clearly errors and the build was reported as failed. Looking at some other jobs it seems the test suite passed at least for some targets even though the status was not reported back to this PR. 🤔

I have fixed all linter errors, rebased again from master and made sure the test suite still passes locally. Apologies for the hurdle!

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10750/

robertrossmann commented:
Looks like the CI did not properly report back to this PR - I randomly checked some of the tasks and they are listed as successful on CI. Just FYI.

Fishrock123 commented on test/parallel/test-signal-args.js:
```diff
@@ -0,0 +1,20 @@
+
+// Prevent Node.js from exiting due to empty event loop before signal handlers
+// are fired
+setTimeout(() => {}, 0);
```
`setImmediate` would be better. Note also that `0` coerces to `1` in timers durations.

Fishrock123 commented:
The CI isn't outputting a stack for some reason:
```
not ok 336 parallel/test-signal-args
  ---
  duration_ms: 0.120
  severity: fail
  stack: |-
  ...
```

Maybe someone from @nodejs/platform-windows can run the test on a local machine?

robertrossmann commented:
@Fishrock123 Thanks for the suggestion, I updated the code.

About windows failure, from the docs:

> Note: Windows does not support sending signals, but Node.js offers some emulation with process.kill(), and subprocess.kill(). Sending signal 0 can be used to test for the existence of a process. Sending SIGINT, SIGTERM, and SIGKILL cause the unconditional termination of the target process.

Could it be that the windows process is being unconditionally terminated due to it receiving `SIGINT` or `SIGTERM`? If that's the case, it would explain why there is no stack trace.

There is a mention of some signal emulation on Windows but I have no experience with win32 systems. Just a wild guess here.

bzoz commented:
The test works for me locally. I see that there were some changes since the CI run, so I've started the job again: https://ci.nodejs.org/job/node-test-pull-request/10772/

bzoz commented:
Ok, after closer look the test fails for me locally to.

Anyhow this test will not work on Windows - catching `SIGTERM` or `SIGKILL` is not supported in libuv and  `process.kill` uses `TerminateProcess`  which can't be trapped. This test should be skipped on Windows.

On a side note - I've tested this manually with Ctrl-C and `SIGINT` was caught and reported as expected on Windows.

robertrossmann commented:
@bzoz I think we could drop the `SIGTERM` part from the test, would that solve the Windows issue?

bzoz commented:
No, `SIGINT` will be detected only if it is made through Ctrl-C. Using `process.kill` to send signals does not work on Windows.

robertrossmann commented:
Thanks @bzoz for your help. I have updated the test to be skipped on Windows.
@Fishrock123 barring no other issues, I think this is ready for final review and CI run.

gireeshpunathil commented:
@Fishrock123 - Windows is skipped from the test, so PTAL
5th (hopefully last) CI: https://ci.nodejs.org/job/node-test-pull-request/10806/

bzoz commented on test/parallel/test-signal-args.js:
```diff
@@ -0,0 +1,24 @@
+const assert = require('assert');
+
+if (common.isWindows) {
+  common.skip('Signal handlers are not supported on Windows');
```
a nit: They are somewhat supported. Maybe "using process.kill to send signals is not supported"?

robertrossmann replied:
Thanks for the suggestion, I updated the comment.

jasnell commented:
@Fishrock123 ... how are you feeling about this PR now?

robertrossmann commented:
Is there anything I can do to have this merged? There is only one pending review from @Fishrock123 which has been addressed (win32 failures). Thank you!

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11138/

BridgeAR commented:
New CI https://ci.nodejs.org/job/node-test-pull-request/11639/

jasnell commented:
This is looking good, just waiting for the last few CI bots to finish

jasnell commented:
Landed in 9b2cf1c

MylesBorins commented:
Is this something we would want to backport to 8.x?
